formats/json/commonMain/src/kotlinx/serialization/json/internal/WriteMode.kt

```diff
@@ -32,8 +32,19 @@ internal inline fun <T, R1 : T, R2 : T> Json.selectMapMode(
     ifMap: () -> R1,
     ifList: () -> R2
 ): T {
-    val keyDescriptor = mapDescriptor.getElementDescriptor(0).carrierDescriptor
-    val keyKind = keyDescriptor.kind
+    var keyDescriptor = mapDescriptor.getElementDescriptor(0).carrierDescriptor
+    var keyKind = keyDescriptor.kind
+
+    if (keyKind == SerialKind.CONTEXTUAL) {
+        keyDescriptor = serializersModule.getContextualDescriptor(keyDescriptor)
+            ?: throw InvalidKeyKindException(keyDescriptor)
+        keyKind = if (keyDescriptor.isInline) {
```

Comment:
That's the same unwrapping that the `.carrierDescriptor` does. I think it all can be merged in the same function like

```kotlin
internal fun SerialDescriptor.carrierDescriptor(module: SerializersModule): SerialDescriptor =
    when {
        kind == SerialKind.CONTEXTUAL -> module.getContextualDescriptor(this)?.carrierDescriptor(module) ?: this
        isInline -> getElementDescriptor(0).carrierDescriptor(module)
        else -> this
    }
```